File: Python/3174. Clear Digits.py
class Solution:
    def clearDigits(self, s: str) -> str:
        # table = {}

        # for i, c in enumerate(s):
        #     table[i] = c.isalpha()

        # for i, c in enumerate(s):
        #     if not c.isdigit():
        #         continue

        #     for j in range(i - 1, -1, -1):
        #         if table[j]:
        #             table[j] = False
        #             break

        # result = ""

        # for i, c in enumerate(s):
        #     if table[i]:
        #         result += c

        # return result

        result = []

        for c in s:
            if c.isalpha():
                result.append(c)
            else:result.pop()

        return "".join(result)

File: C/3174. Clear Digits.c
bool isAlpha(char c) {
    return 97 <= (int)c && (int)c <= 122;
}

void append(char** s, char c) {
    size_t len = strlen(*s);
    char* new = realloc(*s, len + 2);

    new[len] = c;
    new[len + 1] = '\0';

    *s = new;
}

void pop(char** s) {
    size_t len = strlen(*s);

    if (len == 0)
        return;

    char* new = realloc(*s, len);

    new[len - 1] = '\0';
    *s = new;
}

char* clearDigits(const char* s) {
    char* r = malloc(1);

    *r = '\0';

    while (*s) {
        if (isAlpha(*s))
            append(&r, *s);
        else
            pop(&r);

        s++;
    }

    return r;
}
